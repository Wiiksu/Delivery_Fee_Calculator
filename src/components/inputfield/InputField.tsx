import React, { forwardRef, useState } from "react";
import { InputFieldProps } from "../../helpers/types";

export const InputField = forwardRef<HTMLInputElement, InputFieldProps>(
  (
    { label, onChange, defaultValue, errorMessage, ...props }: InputFieldProps,
    ref
  ) => {
    const [fieldValue, setFieldValue] = useState(!!defaultValue);

    const handleChange = (e: React.FocusEvent<HTMLInputElement>) => {
      setFieldValue(e.target.value.length !== 0);
      if (onChange) {
        onChange(e);
      }
    };

    return (
      <section>
        <div className="relative flex w-full items-center">
          <label className="text-purple-50 font-semibold w-[6rem] md:w-40 leading-5 md:leading-none text-base md:text-lg">
            {label}:
          </label>
          <input
            className="w-52 rounded py-1 px-2 outline-none box-border border-collapse border-2 border-purple-50 border-opacity-60 focus:border-purple-500 focus:border-2"
            ref={ref}
            onChange={handleChange}
            data-testid={"inputField"}
            {...props}
          />
        </div>
        {errorMessage ? (
          <p
            data-testid={`${
              props.name ? `${props.name}ErrorMessage` : `errorMessage`
            }`}
            className="pl-[5.7rem] md:pl-40 pt-1 text-sm text-purple-50"
          >
            {errorMessage}
          </p>
        ) : null}
      </section>
    );
  }
);
