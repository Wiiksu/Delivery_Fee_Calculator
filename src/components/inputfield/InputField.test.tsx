import { expect, it, describe, vi } from "vitest";
import { render, screen, fireEvent } from "@testing-library/react";
import React from "react";
import { InputField } from "./InputField";

describe("<InputField />", () => {
  it("Should render correctly to the snapshot", () => {
    const { asFragment } = render(<InputField label="Test Label" />);
    expect(asFragment()).toMatchSnapshot();
  });

  it("Should update value on change", () => {
    render(<InputField label="Test Label" />);
    const input = screen.getByTestId("inputField") as HTMLInputElement;
    fireEvent.change(input, { target: { value: "New Value" } });
    expect(input.value).toBe("New Value");
  });

  it("Should call onChange handler", () => {
    const handleChange = vi.fn();
    render(<InputField label="Test Label" onChange={handleChange} />);
    const input = screen.getByTestId("inputField") as HTMLInputElement;
    fireEvent.change(input, { target: { value: "New Value" } });
    expect(handleChange).toHaveBeenCalled();
  });

  it("Should display error-message", () => {
    const errorMessage = "Error message";
    render(<InputField label="Test Label" errorMessage={errorMessage} />);
    expect(screen.getByText(errorMessage)).toBeInTheDocument();
  });

  it("Should forward ref", () => {
    const ref = React.createRef<HTMLInputElement>();
    render(<InputField label="Test Label" ref={ref} />);
    expect(ref.current).toBeInstanceOf(HTMLInputElement);
  });
});
