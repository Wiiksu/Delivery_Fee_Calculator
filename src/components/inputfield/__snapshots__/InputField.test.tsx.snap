// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`<InputField /> > Should render correctly to the snapshot 1`] = `
<DocumentFragment>
  <section>
    <div
      class="relative flex w-full items-center"
    >
      <label
        class="text-purple-50 font-semibold w-[6rem] md:w-40 leading-5 md:leading-none text-base md:text-lg"
      >
        Test Label:
      </label>
      <input
        class="w-52 rounded py-1 px-2 outline-none box-border border-collapse border-2 border-purple-50 border-opacity-60 focus:border-purple-500 focus:border-2"
        data-testid="inputField"
      />
    </div>
  </section>
</DocumentFragment>
`;
